OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs

```diff
@@ -147,28 +191,45 @@ public override Activity Tick(Actor self)
 				}
 			}
 
-			List<Activity> landingProcedures = new List<Activity>();
+			var exit = dest.FirstExitOrDefault(null);
+			var offset = exit != null ? exit.Info.SpawnOffset : WVec.Zero;
 
-			var turnRadius = Fly.CalculateTurnRadius(aircraft.Info.Speed, aircraft.Info.TurnSpeed);
+			if (aircraft.Info.CanHover)
+				QueueChild(self, new HeliFly(self, Target.FromPos(dest.CenterPosition + offset)), true);
+			else if (aircraft.Info.VTOL)
+				QueueChild(self, new Fly(self, Target.FromPos(dest.CenterPosition + offset)), true);
+			else
+			{
+				var turnRadius = Fly.CalculateTurnRadius(aircraft.Info.Speed, aircraft.Info.TurnSpeed);
 
-			landingProcedures.Add(new Fly(self, Target.FromPos(w1), WDist.Zero, new WDist(turnRadius * 3)));
-			landingProcedures.Add(new Fly(self, Target.FromPos(w2)));
+				QueueChild(self, new Fly(self, Target.FromPos(w1), WDist.Zero, new WDist(turnRadius * 3)), true);
+				QueueChild(self, new Fly(self, Target.FromPos(w2)), true);
 
-			// Fix a problem when the airplane is send to resupply near the airport
-			landingProcedures.Add(new Fly(self, Target.FromPos(w3), WDist.Zero, new WDist(turnRadius / 2)));
+				// Fix a problem when the airplane is sent to resupply near the airport
+				QueueChild(self, new Fly(self, Target.FromPos(w3), WDist.Zero, new WDist(turnRadius / 2)), true);
+			}
 
 			if (ShouldLandAtBuilding(self, dest))
 			{
 				aircraft.MakeReservation(dest);
 
-				landingProcedures.Add(new Land(self, Target.FromActor(dest)));
-				landingProcedures.Add(new ResupplyAircraft(self));
+				if (aircraft.Info.VTOL)
+				{
+					if (aircraft.Info.TurnToDock)
+						QueueChild(self, new Turn(self, aircraft.Info.InitialFacing), true);
+
+					QueueChild(self, new HeliLand(self, false), true);
+				}
+				else
+					QueueChild(self, new Land(self, Target.FromPos(dest.CenterPosition + offset)), true);
+
+				QueueChild(self, new ResupplyAircraft(self), true);
 			}
 
 			if (!abortOnResupply)
-				landingProcedures.Add(NextActivity);
+				QueueChild(self, NextActivity, true);
```

Comment:
Hmm, I don't think it is a good idea to queue `NextActivity` as a childactivity. There should probably just be a `return NextActivity` here somewhere.

Reply:
I moved the abortOnResupply consideration as well as the `return NextActivity` into the `if (ChildActivity != null)`, more precisely both of these are now triggered when the ChildActivity *becomes* `null`.